pyfixest/estimation/feiv_.py

```diff
@@ -193,3 +206,263 @@ def get_fit(self) -> None:
+
+        Example
+        -------
+        The following is an example usage of this method:
+
+            ```{python}
+
+            import numpy as np
+            import pandas as pd
+            from pyfixest.estimation.estimation import feols
+
+            # Set random seed for reproducibility
+            np.random.seed(1)
+
+            # Number of observations
+            n = 1000
+
+            # Simulate the data
+            # Instrumental variable
+            z = np.random.binomial(1, 0.5, size=n)
+            z2 = np.random.binomial(1, 0.5, size=n)
+
+            # Endogenous variable
+            d = 0.5 * z + 1.5 * z2 + np.random.normal(size=n)
+
+            # Control variables
+            c1 = np.random.normal(size=n)
+            c2 = np.random.normal(size=n)
+
+            # Outcome variable
+            y = 1.0 + 1.5 * d + 0.8 * c1 + 0.5 * c2 + np.random.normal(size=n)
+
+            # Cluster variable
+            cluster = np.random.randint(1, 50, size=n)
+            weights = np.random.uniform(1, 3, size=n)
+
+            # Create a DataFrame
+            data = pd.DataFrame({
+                'd': d,
+                'y': y,
+                'z': z,
+                'z2': z2,
+                'c1': c1,
+                'c2': c2,
+                'cluster': cluster,
+                'weights': weights
+            })
+
+            vcov_detail = "iid"
+
+            # Fit OLS model
+            fit_ols = feols("y ~ 1 + d + c1 + c2", data=data, vcov=vcov_detail)
+
+            # Fit IV model
+            fit_iv = feols("y ~ 1 + c1 + c2 | d ~ z", data=data,
+                     vcov=vcov_detail,
+                     weights="weights")
+            fit_iv.first_stage()
+            F_stat_pf = fit_iv._f_stat_1st_stage
+            fit_iv.IV_Diag()
+            F_stat_eff_pf = fit_iv._eff_F
+
+            print("(Unadjusted) F stat :", F_stat_pf)
+            print("Effective F stat :", F_stat_eff_pf)
+
+            # The example above generates the following results
+            # (Unadjusted) F stat : 52.81535560457482
+            # Effective F stat : 48.661542741328205
+        """
+        iv_diag_statistics = iv_diag_statistics or []
+
+        if "f_stat" in iv_diag_statistics:
+            self._p_iv = len(self._non_exo_instruments)
+
+            # Create an identity matrix of size p_iv by p_iv
+            # Pad the identity matrix with zeros to make it of size p_iv by k
+            p_iv = self._p_iv  # number of IVs
+            k = (
+                self._model_1st_stage._k
+            )  # number of estimated coefficients of 1st stage
+
+            # Generate matrix R that tests the following;
+            # H0 : \beta_{z_1} = 0 & ... & \beta_{z_{p_iv}} = 0
+            #      where z_1, ..., z_{p_iv} are the instrument variables
+            # H1 : H0 does not hold
+
+            identity_matrix = np.eye(p_iv)
+            R = np.zeros((p_iv, k))
+            if self._has_fixef:
```

Comment:
Here we implicitly encode that the order of the first stage design matrix is [intercept, instruments, exogenous], right? I am afraid this might be "unsafe" in case this order was ever to change. Have been thinking about this about the last 5 minutes but have no great answer.

Reply:
Could we maybe do something like 

```python
%load_ext autoreload
%autoreload 2
import pyfixest as pf 
import numpy as np

data = pf.get_data()

fit = pf.feols("X1 ~ Z1", data = data)

k = fit._k 
p_iv = 1

identity_matrix = np.eye(p_iv)
R = np.zeros((p_iv, k))

# those we might store in the Feiv object? 
instruments = ["Z1"]
idx = [fit._coefnames.index(x) for x in instruments if x in fit._coefnames]
idx

R[:, 1 : p_iv + 1] = identity_matrix
R
# array([[0., 1.]])
```
?

Reply:
Ah, I think that we should actually have what I call "instruments in `self._non_exo_instruments` above, or not?

Reply:
Yes, you are right. I think there are two points that implicitly assume the position of instruments 

```python 

....
            if self._has_fixef:
                R[:, :p_iv] = identity_matrix
            else:
                R[:, 1 : p_iv + 1] = identity_matrix
....

           iv_pos = range(0, self._p_iv) if self._has_fixef else range(1, self._p_iv + 1)


```

This will be problematic if there is any chance that the positions of IVs changes, so my strategy will be 1) extract the first index of IVs from the 1st stage regression and the number of IVs (i.e. `_p_iv`), 2) and make range object using this to locate identity matrix and and extract submatrix of `vcov`.

The change code will be like the following : 

```python
....
    def extract_indices_IVs(self):
        """Extract indices of instrument variales in the first stage regression

        Returns
        -------
        None

        """
                
        indices = []
        for instrument in self._non_exo_instruments:
            if instrument in self._coefnames_z:
                index = self._coefnames_z.index(instrument)
                indices.append(index)
        return indices

....

       self._iv_loc_first = np.min(iv_loc)
       iv_loc_first = self._iv_loc_first
       R[:,iv_loc_first:p_iv+iv_loc_first] = identity_matrix

....

        # Extract the submatrix
        iv_pos = range(self._iv_loc_first, self._p_iv+self._iv_loc_first)
```